### DISCUSSION TITLE
Distinguish reINVITE

### DISCUSSION THREAD

**Discussion by pbalash0v** [resiprocate/resiprocate#271] [Q&A] [answered]
Hello,

What would be preferred way to distinguish reINVITE-triggered invocation of `resip::InviteSessionHandler`'s derived class's `onOffer` callback from initial INVITE-triggered call of the same callback ?
And (almost) similar question is for distinguishing reINVITE-triggered invocations for `onOfferRejected` and `onAnswer`.

So far I've came up with relatively hackish (imo) way to keep a `resip::DialogEventInfo::State` field in my `resip::AppDialog`'s derived instance (gets updated in `resip::DialogEventHandler`'s derived class callbacks). Then I can aquire this state value in `onOffer` from provided `resip::InviteSessionHandle` parameter and check against `resip::DialogEventInfo::Confirmed`. But it feels cumbersome.

Thanks.

**Comment by sgodin** [comment, marked as answer]
I have never had a need to do such a thing in the many, many projects I've done over the years with DUM.  DUM keeps the SDP offer/answer state machine decoupled from the INVITE state machine.   For inbound/server side processing, this means you shouldn't be processing SDP in the onNewSession callback, you should wait for the onOffer callback.

That being said, if you really need to do this for some reason there are a few approaches you could use:

- Use an approach like basicClient uses to have a basicClientCall object for each InviteSession, and in this class just store/maintain a flag to track initial offer vs mid-dialog offer
- the onOffer callback gives you the full SIP message that triggered it.  On your first inbound INVITE the to tag (To header) will not be set.  It will be set on all mid-dialog INVITEs though.
- use the InviteSessionHandle to access InviteSession::isConnected method to query if the session is already connected or not - this should be equivalent to what you did in a round about way with the DialogEventHandler, but is much more straight forward

Scott Godin
http://www.sipspectrum.com/

**Reply by pbalash0v** [reply]
Scott, thanks a lot, really appreciated. Will definitely check your suggestions.

Pavel.
